src/rank_llm/retrieve/service_retriever.py

```diff
@@ -0,0 +1,120 @@
+from enum import Enum
+from typing import Any, Dict, List, Union
+
+from rank_llm.data import Request, Candidate, Query
+from rank_llm.retrieve.pyserini_retriever import PyseriniRetriever, RetrievalMethod
+from rank_llm.retrieve.repo_info import HITS_INFO
+from rank_llm.retrieve.utils import compute_md5, download_cached_hits
+
+
+class RetrievalMode(Enum):
+    DATASET = "dataset"
+    CUSTOM = "custom"
+
+    def __str__(self):
+        return self.value
+
+
+class ServiceRetriever:
+    def __init__(
+        self,
+        dataset: Union[str, List[str], List[Dict[str, Any]]],
+        host: str,
+        retrieval_mode: RetrievalMode = RetrievalMode.DATASET,
+        retrieval_method: RetrievalMethod = RetrievalMethod.BM25,
+        # index_path: str = None,
+        # topics_path: str = None,
+        # index_type: str = None,
+        # encoder: str = None,
+        # onnx: bool = False,
+    ) -> None:
+        self._retrieval_mode = retrieval_mode
+        self._host = host
+        self._retrieval_method = retrieval_method
+        self._dataset = dataset
+        # self._index_path = index_path
+        # self._topics_path = topics_path
+        # self._index_type = index_type
+        # self._encoder = encoder
+        # self._onnx = onnx
+
+    @staticmethod
+    def from_dataset_with_prebuilt_index(
+        request: Request, 
+        dataset_name: str,
+        retrieval_method: RetrievalMethod = RetrievalMethod.BM25,
+        k: int = 100, #Anserini API currently does not support passing in k
+        host: str = "http://localhost:8081",
+    ):
+        """
+        Creates a Retriever instance for a dataset with a prebuilt index.
+
+        Args:
+            dataset_name (str): The name of the dataset.
+            retrieval_method (RetrievalMethod): The retrieval method to be used. Defaults to BM25.
+            k (int, optional): The top k hits to retrieve. Defaults to 100.
+
+        Returns:
+            Request. Contains a query and list of candidates
+        Raises:
+            ValueError: If dataset name or retrieval method is invalid or missing.
+        """
+        if not dataset_name:
+            raise ValueError("Please provide name of the dataset.")
+        if not isinstance(dataset_name, str):
+            raise ValueError(
+                f"Invalid dataset format: {dataset_name}. Expected a string representing name of the dataset."
+            )
+        if not retrieval_method:
+            raise "Please provide a retrieval method."
+        if retrieval_method == RetrievalMethod.UNSPECIFIED:
+            raise ValueError(
+                f"Invalid retrieval method: {retrieval_method}. Please provide a specific retrieval method."
+            )
+        retriever = ServiceRetriever(
+            retrieval_mode=RetrievalMode.DATASET,
+            retrieval_method=retrieval_method,
+            dataset=dataset_name,
+            host=host,
+        )
+        return retriever.retrieve(k=k, request=request)
```

Comment:
please return the retriever, the user can reuse the retriever for multiple retrieve requests rather than instantiating retriever every single time.  This way k and request args could be passed to retrieve function rather than init.